### ISSUE TITLE
',' in search param is mistreated as '\'

### ISSUE THREAD

**Issue by SherryH** [jamesagnew/hapi-fhir#192]
In performing CarePlan search, if a comma ',' is contained in Token search param, the search result returns contains backslash '\' instead if the search code is escaped according to [FHIR spec](http://www.hl7.org/fhir/search.html#escaping). Empty list is returned if it is not escaped.

In string search without escaping, ',' is ignored and all the partial matching strings are returned.

**Comment by jamesagnew** [comment]
Hi @SherryH ,

I'm not able to reproduce this so far, although I"m probably missing something. Would you be able to share a code sample illustrating the issue?

Cheers,
James

**Comment by SherryH** [comment]
Hi @jamesagnew ,
Below is an extraction of the code. *,* is treated as *\*  when used at the end of the search parameter.

Thanks,
Sherry
------------------------------------
@Search()
 public List<Practitioner> getResourcesByIdentifiers(
   @RequiredParam(name = Practitioner.SP_IDENTIFIER) final TokenOrListParam theIdentifiers) {

  // the special chars: ',', '\' is not properly handled in extreme case
  // e.g.
  // -- RIGHT -- "Practitioner?identifier=system|code-include-but-not-end-with-comma\,suffix" is solved as system = "system", and code= "code-include-but-not-end-with-comma,suffix"
  // -- WRONG -- "Practitioner?identifier=system|code-end-with-comma\," is solved as system = "system", and code= "code-end-with-comma\"

 }

**Comment by jamesagnew** [comment]
Hi Sherry- I've added a unit test that tests exactly this, and it still seems to work fine (see https://github.com/jamesagnew/hapi-fhir/commit/af0db664db86a0596af1aa09a5b1f8a19dc37bbf ).

Is it possible that there is an issue on the client end? E.g. "\," is getting converted to ","? Are you using HAPI's client, or something else?

**Comment by jamesagnew** [comment]
Hi Sherry, after your comment on the last commit (thanks for that!) I was able to reproduce.

This is fixed, and in the process I also discovered that we also aren't always correctly handling escaped backslashes (```foo\\bar``` should unescape to ```foo\bar```) so this is fixed too. I'll push a git fix shortly, and there should be a new 1.2-SNAPSHOT build up in Maven within the next day or so.

Thanks again!!
